### PR TITLE
[MPS] Implement FFT operations

### LINKED ISSUE #78044: FFT operators are not supported on MPS device
### 🐛 Describe the bug

# Extra comments

The error happens regardless of whether `PYTORCH_ENABLE_MPS_FALLBACK=1` env var is set or not.

# Code to reproduce:

```python
import torch
x = torch.randn(1, 16000, device="mps")
y = torch.fft.rfft(x)
y_abs = y.abs()
```

# Script output message:

```
$ python test2.py
test2.py:4: UserWarning: The operator 'aten::_fft_r2c' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  y = torch.fft.rfft(x)
libc++abi: terminating with uncaught exception of type c10::TypeError: Trying to convert ComplexFloat to the MPS backend but it does not have support for that dtype.
Exception raised from getMPSDataType at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/native/mps/OperationUtils.mm:124 (most recent call first):
frame #0: at::native::mps::getMPSDataType(c10::ScalarType) + 452 (0x113dc690c in libtorch_cpu.dylib)
frame #1: at::native::mps::mpsGraphRankedPlaceHolder(MPSGraph*, at::Tensor const&) + 60 (0x113dc838c in libtorch_cpu.dylib)
frame #2: invocation function for block in at::native::mps::unary_op(at::Tensor const&, at::Tensor const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, MPSGraphTensor* (MPSGraph*, MPSGraphTensor*) block_pointer) + 96 (0x113e33d84 in libtorch_cpu.dylib)
frame #3: invocation function for block in at::native::mps::MPSGraphCache::CreateCachedGraph(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, at::native::mps::MPSCachedGraph* () block_pointer) + 216 (0x113dd0dec in libtorch_cpu.dylib)
frame #4: _dispatch_client_callout + 20 (0x1984d01b4 in libdispatch.dylib)
frame #5: _dispatch_lane_barrier_sync_invoke_and_complete + 56 (0x1984df414 in libdispatch.dylib)
frame #6: at::native::mps::MPSGraphCache::CreateCachedGraph(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, at::native::mps::MPSCachedGraph* () block_pointer) + 160 (0x113dc8e90 in libtorch_cpu.dylib)
frame #7: at::native::mps::unary_op(at::Tensor const&, at::Tensor const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, MPSGraphTensor* (MPSGraph*, MPSGraphTensor*) block_pointer) + 984 (0x113e33b70 in libtorch_cpu.dylib)
frame #8: at::native::abs_out_mps(at::Tensor const&, at::Tensor&) + 76 (0x113e351a4 in libtorch_cpu.dylib)
frame #9: at::_ops::abs_out::call(at::Tensor const&, at::Tensor&) + 280 (0x110f268e0 in libtorch_cpu.dylib)
frame #10: at::native::abs(at::Tensor const&) + 232 (0x110aadcd4 in libtorch_cpu.dylib)
frame #11: c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::DispatchKeySet, at::Tensor const&), &(torch::autograd::VariableType::(anonymous namespace)::abs(c10::DispatchKeySet, at::Tensor const&))>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&> >, at::Tensor (c10::DispatchKeySet, at::Tensor const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&) + 968 (0x1125fa24c in libtorch_cpu.dylib)
frame #12: at::_ops::abs::call(at::Tensor const&) + 264 (0x110f25c84 in libtorch_cpu.dylib)
frame #13: torch::autograd::THPVariable_abs(_object*, _object*) + 188 (0x106de8904 in libtorch_python.dylib)
frame #14: method_vectorcall_NOARGS + 136 (0x104e5ee4c in python3.8)
frame #15: call_function + 128 (0x104f859a0 in python3.8)
frame #16: _PyEval_EvalFrameDefault + 19988 (0x104f7c91c in python3.8)
frame #17: _PyEval_EvalCodeWithName + 616 (0x104f75cc8 in python3.8)
frame #18: pyrun_file + 280 (0x104fea0d0 in python3.8)
frame #19: pyrun_simple_file + 448 (0x104fe980c in python3.8)
frame #20: PyRun_SimpleFileExFlags + 120 (0x104fe95ec in python3.8)
frame #21: pymain_run_file + 444 (0x105011040 in python3.8)
frame #22: pymain_run_python + 328 (0x10501040c in python3.8)
frame #23: Py_RunMain + 40 (0x105010268 in python3.8)
frame #24: pymain_main + 52 (0x105011c50 in python3.8)
frame #25: main + 56 (0x104e25a08 in python3.8)
frame #26: start + 520 (0x10534508c in dyld)

[1]    17751 abort      python test2.py
```

### Versions

# Env information

```
$ python collect_env.py
Collecting environment information...
PyTorch version: 1.13.0.dev20220521
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.22.4
Libc version: N/A

Python version: 3.8.13 (default, Mar 28 2022, 06:13:39)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] k2==1.15.1.dev20220520+cpu.torch1.12.0.dev20220519
[pip3] numpy==1.22.3
[pip3] torch==1.13.0.dev20220521
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] k2                        1.15.1.dev20220520+cpu.torch1.12.0.dev20220519          pypi_0    pypi
[conda] numpy                     1.22.3           py38h25ab29e_0
[conda] numpy-base                1.22.3           py38h974a1f5_0
[conda] torch                     1.13.0.dev20220521          pypi_0    pypi
[conda] torchaudio                0.11.0                   pypi_0    pypi
[conda] torchvision               0.12.0                   pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @peterbell10 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by beveradb**
Fixes #78044 

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

**Comment by agnosticlines** [comment]
Just wanted to say thanks for working on this, needed for an audio processing pipeline I've been working on, very exciting to see the potential speedup compared to running on cpu!

**Comment by beveradb** [comment]
Hey @malfet do you know if I'm missing something obvious re. the CI build failure?

```
/Users/ec2-user/runner/_work/pytorch/pytorch/aten/src/ATen/native/mps/MPSGraphSonomaOps.h:3:10: note: did not find header 'MPSGraphFourierTransformOps.h' in framework 'MetalPerformanceShadersGraph' (loaded from '/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk/System/Library/Frameworks')
```

I opened [this issue](https://github.com/pytorch/test-infra/issues/4899) because I believe it's just that the Github Actions runner is still using an older image with SDK version 13 when it needs SDK version 14 to have the right header, but maybe I'm missing something or that's something I could configure myself?

Apologies, I'm still a bit overwhelmed by the pytorch build/test workflow and infra; I'd hoped to ask for a little advice in the slack community but haven't been added yet.

File: aten/src/ATen/native/mps/operations/FourierTransform.mm [hunk 1]
```diff
@@ -0,0 +1,196 @@
+//  Copyright © 2023 Apple Inc.
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/native/UpSample.h>
+#include <ATen/native/mps/MPSGraphVenturaOps.h>
+#include <ATen/native/mps/MPSGraphSonomaOps.h>
+#include <ATen/native/mps/OperationUtils.h>
+
+#include <MetalPerformanceShadersGraph/MPSGraph.h>
+#include <MetalPerformanceShadersGraph/MPSGraphFourierTransformOps.h>
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/FourierTransform.mm:9]: You don't need this one here, do you (as you redefine ops in MPSGraphSonomaOps, don't you?
```suggestion
```

File: aten/src/ATen/native/mps/MPSGraphSonomaOps.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#pragma once
+#include <MetalPerformanceShadersGraph/MetalPerformanceShadersGraph.h>
+#include <MetalPerformanceShadersGraph/MPSGraphFourierTransformOps.h>
```
Comment (malfet) [inline aten/src/ATen/native/mps/MPSGraphSonomaOps.h:3]: Aren't you redefining those ops here? If so, why do you include the header?
```suggestion
```

**Comment by malfet** [comment]
> Do you know if I'm missing something obvious re. the CI build failure?
> 
> ```
> /Users/ec2-user/runner/_work/pytorch/pytorch/aten/src/ATen/native/mps/MPSGraphSonomaOps.h:3:10: 
> note: did not find header 'MPSGraphFourierTransformOps.h' in framework 'MetalPerformanceShadersGraph' 
> (loaded from '/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk/System/Library/Frameworks')
> ```

Yes, the whole point of MPSGraphSonomaOps.h is not redefine Ops which are yet not available in the base SDK, so deleting those headers should probably fix the compilation issues. 

Another problem you currently have with your PR, are contributions from @sriege which you need to remove, or convince them to sign the CLA

**Comment by sriege** [comment]
@malfet I signed the CLA correctly last week, if it's not linking to my account correctly for some reason I will look into this shortly

**Comment by malfet** [comment]
@sriege can you check that email reported in https://github.com/pytorch/pytorch/commit/7ff3e2e4a4c6f38a3cf99e1e227ead2fdecaffd2.patch is the email your github account is associated with?

**Comment by sriege** [comment]
@malfet we should be good!

File: aten/src/ATen/native/mps/MPSGraphSonomaOps.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#if !defined(__MAC_14_0) && \
+    (!defined(MAC_OS_X_VERSION_14_0) || (MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_14_0))
+
+typedef NS_ENUM(NSUInteger, MPSGraphFFTScalingMode)
+{
+    MPSGraphFFTScalingModeNone          = 0L,
+    MPSGraphFFTScalingModeSize          = 1L,
+    MPSGraphFFTScalingModeUnitary       = 2L,
+};
+#endif
```
Comment (malfet) [inline aten/src/ATen/native/mps/MPSGraphSonomaOps.h:15]:
```suggestion
#if !defined(__MAC_14_0) && \
    (!defined(MAC_OS_X_VERSION_14_0) || (MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_14_0))

typedef NS_ENUM(NSUInteger, MPSGraphFFTScalingMode)
{
    MPSGraphFFTScalingModeNone          = 0L,
    MPSGraphFFTScalingModeSize          = 1L,
    MPSGraphFFTScalingModeUnitary       = 2L,
};

@interface FakeMPSGraphFFTDescriptor : NSObject<NSCopying> 
@property (readwrite, nonatomic) BOOL inverse;
@property (readwrite, nonatomic) MPSGraphFFTScalingMode scalingMode;
@property (readwrite, nonatomic) BOOL roundToOddHermitean;
@end

@compatibility_alias MPSGraphFFTDescriptor FakeMPSGraphFFTDescriptor
#endif

@interface MPSGraph (SonomaOps)
```

File: aten/src/ATen/native/mps/MPSGraphSonomaOps.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+@interface FakeMPSGraphFFTDescriptor : NSObject<NSCopying> 
+@property (readwrite, nonatomic) BOOL inverse;
+@property (readwrite, nonatomic) MPSGraphFFTScalingMode scalingMode;
+@property (readwrite, nonatomic) BOOL roundToOddHermitean;
++(nullable instancetype) descriptor;
+@end
+
+@compatibility_alias MPSGraphFFTDescriptor FakeMPSGraphFFTDescriptor
```
Comment (malfet) [inline aten/src/ATen/native/mps/MPSGraphSonomaOps.h:21]:
```suggestion
@compatibility_alias MPSGraphFFTDescriptor FakeMPSGraphFFTDescriptor;
```

**Comment by beveradb** [comment]
> Yes, the whole point of MPSGraphSonomaOps.h is not redefine Ops which are yet not available in the base SDK, so deleting those headers should probably fix the compilation issues.

Ah, thank you for confirming, that was what I thought originally but honestly I was pretty confused and was kinda scrambling to try and get it to build 😅 

Thanks for jumping in and helping tidy this up! 🙇

File: aten/src/ATen/native/mps/operations/FourierTransform.mm [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_fft_c2c.h>
+#include <ATen/ops/_fft_c2r.h>
+#include <ATen/ops/_fft_r2c.h>
+#endif
+
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/FourierTransform.mm:18]:
```suggestion

#if !defined(__MAC_14_0) && \
    (!defined(MAC_OS_X_VERSION_14_0) || (MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_14_0))
@implementation FakeMPSGraphFFTDescriptor
+(nullable instancetype) descriptor {
  // This should never be called
  return nil;
}
@end
    
#endif    
```

File: aten/src/ATen/native/mps/operations/FourierTransform.mm [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+Tensor _fft_r2c_mps(const Tensor& input, int64_t signal_ndim, bool normalized, IntArrayRef signal_sizes) {
+  return runFFTGraph(input, {}, false, "fft_r2c_mps", mps::FFTType::R2C, signal_ndim, normalized, true);
+}
+
+// Real-to-Hermitean FFT, writing result to the provided output tensor
+Tensor& _fft_r2c_mps_out(const Tensor& input,
+                         Tensor& out,
+                         int64_t signal_ndim,
+                         bool normalized,
+                         IntArrayRef signal_sizes) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: Sorry, but where did you get this function signature from?

Comment (agnosticlines) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: Hey there @beveradb, Just wondering, I saw for your projects you converted to using MLX, is there still interest in working on this? I'd be willing to sponsor a bounty to get this over the line if the original motivation has gone :-)

Comment (BBC-Esq) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: > Hey there @beveradb, Just wondering, I saw for your projects you converted to using MLX, is there still interest in working on this? I'd be willing to sponsor a bounty to get this over the line if the original motivation has gone :-)

Put me in for contributing towards the bounty as well as long as my portion is reasonable.

Comment (agnosticlines) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: I'm not actually sure how to best do the bounty, the platform I used to use seems to have disappeared and taken everyones money, but I'd be willing to contribute 200 USD towards seeing the FFT ops land in pytorch, I think they're the last ops needed for GPU accelerated TTS/Audio manipulation

Comment (BBC-Esq) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: > I'm not actually sure how to best do the bounty, the platform I used to use seems to have disappeared and taken everyones money, but I'd be willing to contribute 200 USD towards seeing the FFT ops land in pytorch, I think they're the last ops needed for GPU accelerated TTS/Audio manipulation

My landlord would shreak if I contributed more than $20...sorry, but hope it'll help.

Comment (beveradb) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: Appreciate the offer guys, the motivation is still there but I've had to prioritize all of my after-work hours towards other paid work recently due to unplanned family/personal life events.
If nobody else beats me to it I will eventually get back to working on this again when I have more free time, but it's honestly mostly a mental blocker of me feeling out of my depth with it which is the bigger issue; if it was easy I'd have done it already.
If you want it to be implemented sooner, find someone who's actually an experienced Objective-C / C++ developer, preferably with PyTorch and/or macOS development experience, as I'm just fumbling and guessing / doing my best with limited understanding!

**Comment by BBC-Esq** [comment]
It's interesting to me that an organization such as Pytorch doesn't have anyone who can tackle this...

File: aten/src/ATen/native/mps/operations/FourierTransform.mm (hunk 6 above)
Comment (agnosticlines) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: That's fair enough! Sorry to hear that @beveradb hope everything works out okay. Yeah it's not a trivial implementation, I'd work on it myself but I have almost zero objective-c experience and minimal pytorch C++ code experience. I really appreciate your efforts, your life and family come first of course. Hope it all works out :-)

Comment (agnosticlines) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: Hey @malfet sorry for the @, just wondering if you'd know anyone who'd be interested in completing this for a bounty? I know it's not a lot compared to the amount of experience required to work on it, but I'm a single person, not a big company or anything

**Comment by malfet** [comment]
Closing this one, as change enabling FFT has been landed

**Comment by BBC-Esq** [comment]
Where has it been implemented?

**Comment by agnosticlines** [comment]
> Closing this one, as change enabling FFT has been landed

Sorry to ask, but can you link the commit? I've tried to find it in the recent commits and can't find anything mentioning it in the commit name/description

File: aten/src/ATen/native/mps/operations/FourierTransform.mm (hunk 6 above)
Comment (malfet) [reply aten/src/ATen/native/mps/operations/FourierTransform.mm:169]: @agnosticlines FFT should be available in nightlies (committed as https://github.com/pytorch/pytorch/pull/119670 ), if something still does not work, please file an issue

**Comment by beveradb** [comment]
It's here @agnosticlines / @BBC-Esq 😄 

https://github.com/pytorch/pytorch/commit/53bfae2c066fcd06784dfa051cd7e2eb5ba5c8fa

**Comment by agnosticlines** [comment]
> @agnosticlines if code is public, feel free to file an issue, will try to take a look. And I remember there were a PR about upsample_1d somewhere, lets see if it could be landed (or if anyone volunteers to implement it, I can be a reviewer)

Yeah the code is public, I'm using it for a game server backend to do announcer voices, it's alltalk_tts https://github.com/erew123/alltalk_tts, It doesn't support metal natively but I've just flipped all the "cpu" to "mps", maybe a naive approach. That'd be awesome, I really appreciate all your support in this thread :)

**Comment by alanngnet** [comment]
noob question: when will this commit hit the main torch distribution? I'm on torch 2.2.0 and a simple "pip install --upgrade torch" doesn't seem to get me this commit. I still get "The operator 'aten::_fft_c2c' is not currently supported on the MPS backend" from torch.fft.fft()

**Comment by agnosticlines** [comment]
> noob question: when will this commit hit the main torch distribution? I'm on torch 2.2.0 and a simple "pip install --upgrade torch" doesn't seem to get me this commit. I still get "The operator 'aten::_fft_c2c' is not currently supported on the MPS backend" from torch.fft.fft()

You want to install the nightly

> pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu

**Comment by agnosticlines** [comment]
@malfet hey, I tried to find the PR and see if it was something I could work on but can't find it, not sure if you were mistaken/misremembering but seems like this has to be implemented from scratch :(

**Comment by malfet** [comment]
@agnosticlines this one https://github.com/pytorch/pytorch/pull/115031 merging now

**Comment by agnosticlines** [comment]
> @agnosticlines this one #115031 merging now

Thanks so much! This is huge! I have one question and it's not really related to this but I've tried the AllTalk_TTS application with swapping out `cpu`/`cuda` for `mps` but all it produces on mps is like, beeping/popping, sounds almost like a sine wave, is there any way to debug what's causing this?

**Comment by malfet** [comment]
@agnosticlines I think best course of actions is to file a separate issue about it (so that more eyes are on the issue)

**Comment by agnosticlines** [comment]
> @agnosticlines I think best course of actions is to file a separate issue about it (so that more eyes might be looking at it)
> 
> Though I've just tried it myself, and looks like both CPU and MPS works (albeit MPS is a bit slower than CPU) using following example:
> 
> ```python
> from TTS.api import TTS
> for device in ["mps", "cpu"]:
>     tts = TTS("tts_models/multilingual/multi-dataset/xtts_v2").to(device)
>     tts.tts_to_file(text="I think best course of actions is to file a separate issue about it",
>                     speaker_wav="female_06.wav", language="en", file_path=f"output-{device}.wav")
> ```

Will do, thanks very much :)

**Comment by solxnp** [comment]
> > 菜鸟问题：这个提交什么时候会到达主火炬发行版？我使用的是 torch 2.2.0，一个简单的“pip install --upgrade torch”似乎并没有让我得到这个提交。我仍然从 torch.fft.fft() 得到“MPS 后端当前不支持运算符 'aten::_fft_c2c'”
> 
> 您想要安装每晚
> 
> > pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu

I tried to test with CPU and MPS separately, and found that MPS took twice as long as CPU. Why is that?

use cpu 
> Processing time: 63.414340019226074
> Real-time factor: 1.7267567789412566

use mps
> Processing time: 118.5271208Z862854
> Real-time factor: 3.5433564370583683

**Comment by solxnp** [comment]
> > noob question: when will this commit hit the main torch distribution? I'm on torch 2.2.0 and a simple "pip install --upgrade torch" doesn't seem to get me this commit. I still get "The operator 'aten::_fft_c2c' is not currently supported on the MPS backend" from torch.fft.fft()
> 
> You want to install the nightly
> 
> > pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu

I tried to test with CPU and MPS separately, and found that MPS took twice as long as CPU. Why is that?

use cpu

 > Processing time: 63.414340019226074
 > Real-time factor: 1.7267567789412566

use mps

 > Processing time: 118.5271208Z862854
 > Real-time factor: 3.5433564370583683

**Pull request closed without merging**